README.md
# hampa
A half-finished game made with Haskell and the FRP library Yampa. The game is inspired by Pong but is instead of ping pong more like volley ball. A lot of the code could be done cleaner and clearer but this project was for me a way to practice Haskell.

ChangeLog.md
# Changelog for hampa

## Unreleased changes


app/Test.hs
module Test where


app/Game.hs
{-# LANGUAGE Arrows, BangPatterns #-}

module Game where

import FRP.Yampa
import FRP.Yampa.Vector2

type Vector = Vector2 Float
type Size = Vector

data Ball = Ball { ballRadius :: Float, ballPos :: Vector, ballVel :: Vector }

x = vector2X :: Vector -> Float
y = vector2Y :: Vector -> Float
gravity = vector2 0 (-9.8 * 50) :: Vector

ball :: Float -> Vector -> Vector -> Ball
ball radius position velocity = Ball { ballRadius = radius, ballPos = position, ballVel = velocity }

createBall :: Ball -> SF a Ball
createBall ball = proc _ -> do
    vel <- (^+^ ballVel ball) ^<< integral -< gravity
    pos <- (^+^ ballPos ball) ^<< integral -< vel
    returnA -< Ball { ballRadius = ballRadius ball, ballPos = pos, ballVel = vel }

manyBalls :: Size -> Ball -> SF a [Ball]
manyBalls s b = parB . take 10 . repeat $ ballInBox s b

ballInBox :: Size -> Ball -> SF a Ball
ballInBox size ball = 
    let sfBall = createBall ball
        bounceEvent = (\b -> checkBounce size b) ^<< sfBall
    in switch (sfBall &&& bounceEvent) (\b -> ballInBox size b)

checkBounce :: Size -> Ball -> Event Ball
checkBounce s b =
    let r = ballRadius b
        p = ballPos b
        v = ballVel b
    in if x p - r < 0   then Event $ ball r (vector2 (0 + r) (y p)) (vector2 (-(x v)) (y v)) else
       if x p + r > x s then Event $ ball r (vector2 (x s - r) (y p)) (vector2 (-(x v)) (y v)) else
       if y p - r < 0   then Event $ ball r (vector2 (x p) (0 + r)) (vector2 (x v) (-(y v))) else
       if y p + r > y s then Event $ ball r (vector2 (x p) (y s - r)) (vector2 (x v) (-(y v))) else NoEvent

app/Rect.hs
module Rect where

import FRP.Yampa.Vector2
import FRP.Yampa.VectorSpace
import Types
import Debug.Trace

data Rect = Rect Position Size deriving (Eq, Show) 

rectPos :: Rect -> Position
rectPos (Rect p _) = p

rectSize :: Rect -> Position
rectSize (Rect _ s) = s

isOverlapping :: Rect -> Rect -> Bool
isOverlapping (Rect p1 s1) (Rect p2 s2) =
    x p1 < x p2 + x s2 &&
    x p1 + x s1 > x p2 &&
    y p1 < y p2 + y s2 &&
    y p1 + y s1 > y p2

overlap :: Rect -> Rect -> Maybe (Vector2 Double)
overlap (Rect p1 s1) (Rect p2 s2) = do
    dx1 <- x p1 <| (x p2 + x s2)
    dx2 <- (x p1 + x s1) >| x p2
    dy1 <- y p1 <| (y p2 + y s2)
    dy2 <- (y p1 + y s1) |> y p2
    return $ vector2 (minAbs dx1 dx2) (minAbs dy1 dy2)
    where
        minAbs :: Double -> Double -> Double
        minAbs x y = if abs x < abs y then x else y

collision :: Rect -> Rect -> Maybe (Vector2 Double)
collision r1 r2 = do
    v <- overlap r1 r2
    return $ if abs (x v) < abs (y v) then vector2 (x v) 0 else vector2 0 (y v)

collide :: Rect -> Rect -> Maybe Rect
collide r1 r2 = do
    v <- collision r1 r2
    let Rect p s = r2
    return $ Rect (p ^+^ v) s

center :: Rect -> Position
center (Rect p s) = p ^+^ (s ^/ 2)

size :: Rect -> Vector2 Double
size (Rect _ s) = s

(~>) :: a -> (a -> b) -> b
(~>) = flip ($)
        
(|<) :: (Ord a, Num a) => a -> a -> Maybe a
x |< y = if x < y then Just (x - y) else Nothing

(|>) :: (Ord a, Num a) => a -> a -> Maybe a
x |> y = if x > y then Just (x - y) else Nothing

(<|) :: (Ord a, Num a) => a -> a -> Maybe a
x <| y = if x < y then Just (y - x) else Nothing

(>|) :: (Ord a, Num a) => a -> a -> Maybe a
x >| y = if x > y then Just (y - x) else Nothing


app/VolleyPong.hs
{-# LANGUAGE Arrows #-}

module VolleyPong where

import Control.Monad
import Data.Bool
import Data.Char
import Data.List
import Data.Maybe
import Data.Monoid
import Debug.Trace
import FRP.Yampa hiding (fromEvent)
import FRP.Yampa.Vector2
import Prelude hiding (Left, Right)
import Types
import Rect

data GameInput = GameInput Bool (Maybe Direction, Maybe Direction) deriving (Eq, Ord)
data Game = Game { gamePaddle1 :: Paddle, gamePaddle2 :: Paddle, gameBall :: Ball, gameScores :: Scores }
data Paddle = Paddle { paddleRect :: Rect }
data Ball = Ball { ballRect :: Rect, ballVelocity :: Velocity } deriving (Eq, Show)
data Direction = Left | Right deriving (Eq, Ord)
type Scores = (Int, Int)
type Net = Rect

ballGravity = vector2 0 (-400)
ballBounceMagnitude = 500
netHeight = 140
fieldWidth = 800 :: Double
paddleSize = vector2 60 15

gameNet :: Net 
gameNet = let (w, h) = (10, 130) in Rect (vector2 (-w / 2) 0) (vector2 w h)

boolSwitch :: a -> Bool -> Maybe a
boolSwitch x b = if b then Just x else Nothing

paddle :: Paddle -> SF (Maybe Direction) Paddle
paddle p0 = proc mdir -> do
    dir <- arr (fromMaybe zeroVector . fmap dirVec) -< mdir
    pos <- (startPos ^+^) ^<< (speed *^) ^<< integral -< dir
    returnA -< Paddle { paddleRect = Rect pos (vector2 width height) }
    where
        startPos = rectPos $ paddleRect p0
        width = x $ rectSize $ paddleRect p0
        height = y $ rectSize $ paddleRect p0
        speed = 500
        dirVec d = case d of Left  -> vector2 (-1) 0
                             Right -> vector2 1 0

reflect :: Vector2 Double -> Vector2 Double -> Vector2 Double
reflect v n = v ^-^ 2 *^ ((dot v n) *^ n)

clampedPaddle :: (Double, Double) -> Paddle -> SF (Maybe Direction) Paddle
clampedPaddle (xMin, xMax) p0 = switch sf (clampedPaddle (xMin, xMax))
    where
        sf :: SF (Maybe Direction) (Paddle, Event Paddle)
        sf = proc mdir -> do
            p <- paddle p0 -< mdir
            e <- arr clampPaddle -< p
            returnA -< (p, e)
            where
                clampPaddle :: Paddle -> Event Paddle
                clampPaddle p' = if px < xMin || (px + x size) > xMax
                    then Event $ Paddle (Rect (vector2 (clamp xMin (xMax - x size) px) py) size)
                    else NoEvent
                    where
                        py = y $ rectPos $ paddleRect p'
                        px = x $ rectPos $ paddleRect p'
                        size = rectSize $ paddleRect p'

fallingBall :: Ball -> SF a Ball
fallingBall startBall = proc _ -> do
    vel <- (^+^ startVel) ^<< integral -< ballGravity
    pos <- (^+^ startPos) ^<< integral -< vel
    returnA -< Ball { ballRect = Rect pos size, ballVelocity = vel }
    where
        startVel = ballVelocity startBall
        (Rect startPos size) = ballRect startBall

ball :: Ball -> SF (Paddle, Paddle) Ball
ball startBall = switch sf ball
    where
        sf :: SF (Paddle, Paddle) (Ball, Event Ball)
        sf = proc (p1, p2) -> do
            b' <- fallingBall startBall -< ()
            returnA -< (b', mergeEvents [ballNetPhysics b' gameNet, checkBallPaddleColl b' p1, checkBallPaddleColl b' p2])
            where
                checkBallPaddleColl :: Ball -> Paddle -> Event Ball
                checkBallPaddleColl b p = maybeToEvent $ ballPaddleBounce b p

ballPaddleBounce :: Ball -> Paddle -> Maybe Ball
ballPaddleBounce b p = fmap newBall $ collide paddleRect' (ballRect b)
    where
        paddleRect' = paddleRect p
        newBall :: Rect -> Ball
        newBall r = Ball { ballRect = r, ballVelocity = vel }
            where
                vel = ballBounceMagnitude *^ (normalize $ vector2 vx 1)
                vx = dx / phw
                phw = pw / 2
                pw = x . size $ paddleRect'
                dx = x $ (center r) ^-^ (center paddleRect')

ballNetPhysics :: Ball -> Net -> Event Ball
ballNetPhysics ball net = maybeToEvent $ fmap onCollision $ collision br net
    where
        br = ballRect ball
        onCollision :: Vector2 Double -> Ball
        onCollision v = Ball (Rect p' s) (reflect (0.4 *^ bv) n)
            where
                p' = v ^+^ bp
                (Rect bp s) = ballRect ball
                bv = ballVelocity ball
                n = normalize v

game :: SF GameInput Game
game = gameSession (0, 0)

gameSession :: Scores -> SF GameInput Game
gameSession (score1, score2) = switch sf (gameSession . incrementScores (score1, score2))
    where
        sf = proc (GameInput restart (mdir1, mdir2)) -> do
            -- r <- tag Nothing ^<< edge -< restart
            p1 <- clampedPaddle (-1000, 0) paddle1Start -< mdir1
            p2 <- clampedPaddle (0, 1000) paddle2Start -< mdir2
            b <- ball initialBall -< (p1, p2)
            ballOut <- (arr (\b -> if checkBallOutside b then Event b else NoEvent)) -< b
            returnA -< (Game { gamePaddle1 = p1, gamePaddle2 = p2, gameBall = b, gameScores = (score1, score2) }, fmap winnerPlayer ballOut)
            where
                paddle1Start = Paddle $ Rect (vector2 (-200 - x paddleSize) 10) paddleSize
                paddle2Start = Paddle $ Rect (vector2 200 10) paddleSize
                checkBallOutside :: Ball -> Bool
                checkBallOutside b = y' < 0
                    where
                        y' = y bp
                        (Rect bp _) = ballRect b
                initialBall = Ball
                    { 
                        ballRect = Rect (vector2 (-10) 400) (vector2 20 20),
                        ballVelocity = vector2 (-200) 0
                    }

incrementScores :: Scores -> Player -> Scores
incrementScores (x, y) (Player 1) = (x + 1, y)
incrementScores (x, y) (Player 2) = (x, y + 1)
incrementScores s _ = s

data Player = Player Int deriving (Show, Eq)

-- assuming the ball is outside
winnerPlayer :: Ball -> Player
winnerPlayer b = if ballInsideField b then side else otherPlayer side
    where
        otherPlayer (Player 1) = Player 2
        otherPlayer (Player 2) = Player 1
        side = ballPlayerSide b

ballPlayerSide :: Ball -> Player
ballPlayerSide b = if bx < 0 then Player 1 else Player 2
    where
        bx = x $ center $ ballRect $ b

ballInsideField :: Ball -> Bool
ballInsideField = not . isOverlapping fieldRect . ballRect
    where
        fieldRect = Rect (vector2 (-fieldWidth / 2) (-1000000)) (vector2 fieldWidth 20000000)


app/VolleyPongGloss.hs
{-# LANGUAGE Arrows #-}

module VolleyPongGloss where

import VolleyPong

import Control.Monad
import Data.Bool
import Data.Char
import Data.List
import Data.Maybe
import Data.Monoid
import Debug.Trace
import FRP.Yampa hiding (fromEvent)
import FRP.Yampa.Vector2
import Graphics.Gloss
import Graphics.Gloss.Interface.FRP.Yampa
import qualified Graphics.Gloss.Interface.Pure.Game as G
import Prelude hiding (Left, Right)
import Types
import Rect

netPicture :: Picture
netPicture = rectPicture gameNet

paddlePicture :: Paddle -> Picture
paddlePicture = rectPicture . paddleRect
            
ballPicture :: Ball -> Picture
ballPicture = rectPicture . ballRect 

gamePicture :: Game -> Picture
gamePicture g = color white $ translate 0 (-250) $ pictures [paddlePicture $ gamePaddle1 g, -- paddle 1
                                                             paddlePicture $ gamePaddle2 g, -- paddle 2
                                                             ballPicture $ gameBall g, -- ball
                                                             scale 0.8 0.8 $ translate (-300) 400 $ text $ show $ fst $ gameScores g, -- score 1
                                                             scale 0.8 0.8 $ translate (300) 400 $ text $ show $ snd $ gameScores g, -- score 2
                                                             gameFieldPicture,
                                                             netPicture]

gameFieldPicture :: Picture
gameFieldPicture = line [leftCorner, rightCorner]
    where
        leftCorner = (realToFrac $ -fieldWidth / 2, 0)
        rightCorner = (realToFrac $ fieldWidth / 2, 0)

rectPicture :: Rect -> Picture
rectPicture (Rect position size) = translate (x' + w' / 2) (y' + h' / 2) $ rectangleSolid w' h'
    where
        x' = realToFrac $ x position
        y' = realToFrac $ y position
        w' = realToFrac $ x size
        h' = realToFrac $ y size

keyDown :: Char -> SF (Event InputEvent) Bool
keyDown c =
    let f :: Char -> InputEvent -> Event Bool
        f c (G.EventKey (G.Char c') s _ _) = if c == c' then Event (s == G.Down) else NoEvent
        f c _ = NoEvent
    in hold False <<< arr (join . fmap (f c))

inputDirection :: (Char, Char) -> SF (Event InputEvent) (Maybe Direction)
inputDirection (l, r) =
    let inputs = [boolSwitch Left ^<< keyDown l,
                    boolSwitch Right ^<< keyDown r]
    in join . find isJust ^<< parB inputs

inputRestart :: SF (Event InputEvent) Bool
inputRestart = keyDown 'r'

inputPause :: SF (Event InputEvent) Bool
inputPause = keyDown 'p'

appl :: SF (Event InputEvent) Picture
appl = gamePicture ^<< pause game0 inputPause (game <<< input)
    where
        game0 = undefined
        input :: SF (Event InputEvent) GameInput
        input = proc i -> do
            d1 <- inputDirection ('a', 'd') -< i
            d2 <- inputDirection ('j', 'l') -< i
            r <- inputRestart -< i
            returnA -< GameInput r (d1, d2)
    

app/Testing.hs
{-# LANGUAGE Arrows #-}

module Testing where

import Control.Monad
import Data.Bool
import Data.Char
import Data.List
import Data.Maybe
import Data.Monoid
import Debug.Trace
import FRP.Yampa hiding (fromEvent)
import FRP.Yampa.Vector2
import Graphics.Gloss
import Graphics.Gloss.Interface.FRP.Yampa
import qualified Graphics.Gloss.Interface.Pure.Game as G
import Prelude hiding (Left, Right)
import Types
import Rect

data GameInput = GameInput Bool (Maybe Direction, Maybe Direction) deriving (Eq, Ord)
data Game = Game { gamePaddle1 :: Paddle, gamePaddle2 :: Paddle, gameBall :: Ball, gameScores :: Scores }
data Paddle = Paddle { paddleRect :: Rect }
data Ball = Ball { ballRect :: Rect, ballVelocity :: Velocity } deriving (Eq, Show)
data Direction = Left | Right deriving (Eq, Ord)
type Scores = (Int, Int)
type Net = Rect

ballGravity = vector2 0 (-400)
ballBounceMagnitude = 500
netHeight = 140
fieldWidth = 800 :: Double
paddleSize = vector2 60 15

gameNet :: Net 
gameNet = let (w, h) = (10, 130) in Rect (vector2 (-w / 2) 0) (vector2 w h)

boolSwitch :: a -> Bool -> Maybe a
boolSwitch x b = if b then Just x else Nothing

paddle :: Paddle -> SF (Maybe Direction) Paddle
paddle p0 = proc mdir -> do
    dir <- arr (fromMaybe zeroVector . fmap dirVec) -< mdir
    pos <- (startPos ^+^) ^<< (speed *^) ^<< integral -< dir
    returnA -< Paddle { paddleRect = Rect pos (vector2 width height) }
    where
        startPos = rectPos $ paddleRect p0
        width = x $ rectSize $ paddleRect p0
        height = y $ rectSize $ paddleRect p0
        speed = 500
        dirVec d = case d of Left -> vector2 (-1) 0
                             Right -> vector2 1 0

reflect :: Vector2 Double -> Vector2 Double -> Vector2 Double
reflect v n = v ^-^ 2 *^ ((dot v n) *^ n)

clampedPaddle :: (Double, Double) -> Paddle -> SF (Maybe Direction) Paddle
clampedPaddle (xMin, xMax) p0 = switch sf (clampedPaddle (xMin, xMax))
    where
        sf :: SF (Maybe Direction) (Paddle, Event Paddle)
        sf = proc mdir -> do
            p <- paddle p0 -< mdir
            e <- arr clampPaddle -< p
            returnA -< (p, e)
            where
                clampPaddle :: Paddle -> Event Paddle
                clampPaddle p' = if px < xMin || (px + x size) > xMax
                    then Event $ Paddle (Rect (vector2 (clamp xMin (xMax - x size) px) py) size)
                    else NoEvent
                    where
                        py = y $ rectPos $ paddleRect p'
                        px = x $ rectPos $ paddleRect p'
                        size = rectSize $ paddleRect p'

fallingBall :: Ball -> SF a Ball
fallingBall startBall = proc _ -> do
    vel <- (^+^ startVel) ^<< integral -< ballGravity
    pos <- (^+^ startPos) ^<< integral -< vel
    returnA -< Ball { ballRect = Rect pos size, ballVelocity = vel }
    where
        startVel = ballVelocity startBall
        (Rect startPos size) = ballRect startBall

ball :: Ball -> SF (Paddle, Paddle) Ball
ball startBall = switch sf ball
    where
        sf :: SF (Paddle, Paddle) (Ball, Event Ball)
        sf = proc (p1, p2) -> do
            b' <- fallingBall startBall -< ()
            returnA -< (b', mergeEvents [ballNetPhysics b' gameNet, checkBallPaddleColl b' p1, checkBallPaddleColl b' p2])
            where
                checkBallPaddleColl :: Ball -> Paddle -> Event Ball
                checkBallPaddleColl b p = maybeToEvent $ ballPaddleBounce b p

ballPaddleBounce :: Ball -> Paddle -> Maybe Ball
ballPaddleBounce b p = fmap newBall $ collide paddleRect' (ballRect b)
    where
        paddleRect' = paddleRect p
        newBall :: Rect -> Ball
        newBall r = Ball { ballRect = r, ballVelocity = vel }
            where
                vel = ballBounceMagnitude *^ (normalize $ vector2 vx 1)
                vx = dx / phw
                phw = pw / 2
                pw = x . size $ paddleRect'
                dx = x $ (center r) ^-^ (center paddleRect')

ballNetPhysics :: Ball -> Net -> Event Ball
ballNetPhysics ball net = maybeToEvent $ fmap onCollision $ collision br net
    where
        br = ballRect ball
        onCollision :: Vector2 Double -> Ball
        onCollision v = Ball (Rect p' s) (reflect (0.4 *^ bv) n)
            where
                p' = v ^+^ bp
                (Rect bp s) = ballRect ball
                bv = ballVelocity ball
                n = normalize v

game :: SF GameInput Game
game = gameSession (0, 0)

gameSession :: Scores -> SF GameInput Game
gameSession (score1, score2) = switch sf (gameSession . incrementScores (score1, score2))
    where
        sf = proc (GameInput restart (mdir1, mdir2)) -> do
            -- r <- tag Nothing ^<< edge -< restart
            p1 <- clampedPaddle (-1000, 0) paddle1Start -< mdir1
            p2 <- clampedPaddle (0, 1000) paddle2Start -< mdir2
            b <- ball initialBall -< (p1, p2)
            ballOut <- (arr (\b -> if checkBallOutside b then Event b else NoEvent)) -< b
            returnA -< (Game { gamePaddle1 = p1, gamePaddle2 = p2, gameBall = b, gameScores = (score1, score2) }, fmap winnerPlayer ballOut)
            where
                paddle1Start = Paddle $ Rect (vector2 (-200 - x paddleSize) 10) paddleSize
                paddle2Start = Paddle $ Rect (vector2 200 10) paddleSize
                checkBallOutside :: Ball -> Bool
                checkBallOutside b = y' < 0
                    where
                        y' = y bp
                        (Rect bp _) = ballRect b
                initialBall = Ball
                    { 
                        ballRect = Rect (vector2 (-10) 400) (vector2 20 20),
                        ballVelocity = vector2 (-200) 0
                    }

incrementScores :: Scores -> Player -> Scores
incrementScores (x, y) (Player 1) = (x + 1, y)
incrementScores (x, y) (Player 2) = (x, y + 1)
incrementScores s _ = s

data Player = Player Int deriving (Show, Eq)

-- assuming the ball is outside
winnerPlayer :: Ball -> Player
winnerPlayer b = if ballInsideField b then side else otherPlayer side
    where
        otherPlayer (Player 1) = Player 2
        otherPlayer (Player 2) = Player 1
        side = ballPlayerSide b

ballPlayerSide :: Ball -> Player
ballPlayerSide b = if bx < 0 then Player 1 else Player 2
    where
        bx = x $ center $ ballRect $ b

ballInsideField :: Ball -> Bool
ballInsideField = not . isOverlapping fieldRect . ballRect
    where
        fieldRect = Rect (vector2 (-fieldWidth / 2) (-1000000)) (vector2 fieldWidth 20000000)

-- NOT VOLLEYPONG

netPicture :: Picture
netPicture = rectPicture gameNet

paddlePicture :: Paddle -> Picture
paddlePicture = rectPicture . paddleRect
            
ballPicture :: Ball -> Picture
ballPicture = rectPicture . ballRect 

gamePicture :: Game -> Picture
gamePicture g = color white $ translate 0 (-250) $ pictures [paddlePicture $ gamePaddle1 g, -- paddle 1
                                                             paddlePicture $ gamePaddle2 g, -- paddle 2
                                                             ballPicture $ gameBall g, -- ball
                                                             scale 0.8 0.8 $ translate (-300) 400 $ text $ show $ fst $ gameScores g, -- score 1
                                                             scale 0.8 0.8 $ translate (300) 400 $ text $ show $ snd $ gameScores g, -- score 2
                                                             gameFieldPicture,
                                                             netPicture]

gameFieldPicture :: Picture
gameFieldPicture = line [leftCorner, rightCorner]
    where
        leftCorner = (realToFrac $ -fieldWidth / 2, 0)
        rightCorner = (realToFrac $ fieldWidth / 2, 0)

rectPicture :: Rect -> Picture
rectPicture (Rect position size) = translate (x' + w' / 2) (y' + h' / 2) $ rectangleSolid w' h'
    where
        x' = realToFrac $ x position
        y' = realToFrac $ y position
        w' = realToFrac $ x size
        h' = realToFrac $ y size

keyDown :: Char -> SF (Event InputEvent) Bool
keyDown c =
    let f :: Char -> InputEvent -> Event Bool
        f c (G.EventKey (G.Char c') s _ _) = if c == c' then Event (s == G.Down) else NoEvent
        f c _ = NoEvent
    in hold False <<< arr (join . fmap (f c))

inputDirection :: (Char, Char) -> SF (Event InputEvent) (Maybe Direction)
inputDirection (l, r) =
    let inputs = [boolSwitch Left ^<< keyDown l,
                  boolSwitch Right ^<< keyDown r]
    in join . find isJust ^<< parB inputs

inputRestart :: SF (Event InputEvent) Bool
inputRestart = keyDown 'r'

inputPause :: SF (Event InputEvent) Bool
inputPause = keyDown 'p'

appl :: SF (Event InputEvent) Picture
appl = gamePicture ^<< pause game0 inputPause (game <<< input)
    where
        game0 = undefined
        input :: SF (Event InputEvent) GameInput
        input = proc i -> do
            d1 <- inputDirection ('a', 'd') -< i
            d2 <- inputDirection ('j', 'l') -< i
            r <- inputRestart -< i
            returnA -< GameInput r (d1, d2)
